{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/qbio/ari/works/hcmv/hcmv_ont\n"
     ]
    }
   ],
   "source": [
    "%cd /qbio/nest/ari/works/hcmv/hcmv_ont"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from functools import partial\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.rcParams.update({'font.size': 12})\n",
    "matplotlib.rcParams.update({'figure.facecolor': 'white'})\n",
    "\n",
    "matplotlib.rcParams['ytick.labelright'] = True\n",
    "matplotlib.rcParams['ytick.labelleft'] = False\n",
    "matplotlib.rcParams['ytick.right'] = True\n",
    "matplotlib.rcParams['ytick.left'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_anno = pd.read_csv('../refs/HCMV-Toledo.transcripts.bed',sep='\\t',\n",
    "                       names='chr start end name score strand'.split(' '))\n",
    "\n",
    "gene_anno['name'] = [name.lstrip('gene-').lstrip('rna-') for name in gene_anno['name']]\n",
    "\n",
    "target_names = ['RNA1.2', 'RNA2.7', 'RNA4.9', 'UL112', 'UL122',\n",
    "               'UL132', 'UL141', 'UL145', 'UL17', 'UL22A',\n",
    "               'UL25', 'UL34', 'UL4', 'UL40', 'UL48', 'UL48A',\n",
    "               'UL5', 'UL53', 'UL54', 'UL69', 'UL73',\n",
    "               'UL82', 'UL84', 'UL89', 'UL99', 'US1', 'US12', 'US18']\n",
    "target_anno = gene_anno[gene_anno['name'].isin(target_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIN_COV = 50\n",
    "MIN_ODDR = 2\n",
    "AGG_DIS = 6\n",
    "MIN_MODFRAC = 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_names = 'latent_react latent_lytic react_latent react_lytic lytic_latent lytic_react'.split(' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_sequence = '../refs/HCMV-Toledo.fasta'\n",
    "refseq = str(next(SeqIO.parse(reference_sequence, 'fasta')).seq).upper().replace('T', 'U')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_combine_all(filelist, samples):\n",
    "    modfracs = []\n",
    "    for file, name in zip(filelist, samples):\n",
    "        stats = pd.read_csv(file, sep = '\\t')\n",
    "        stats['sample'] = name\n",
    "        modfracs.append(stats)\n",
    "    \n",
    "    modfracs = pd.concat(modfracs).sort_values(by=['start_loc', 'sample'])\n",
    "    return(modfracs)\n",
    "\n",
    "def get_modified_sites(filelist, samples, pvalue=0.05, hp_filter=True):\n",
    "    positive_positions = set()\n",
    "    for file, name in zip(filelist, samples):\n",
    "        stats = pd.read_csv(file, sep = '\\t')\n",
    "        stats['ctrl_reads'] = stats['ctrl_err_1']+stats['ctrl_cor_1']\n",
    "        if hp_filter:\n",
    "            stats = stats[stats['homo_seq']=='--']\n",
    "        stats_high = stats[(stats['total_reads'] >= MIN_COV) &\n",
    "                           (stats['ctrl_reads'] >=MIN_COV) &\n",
    "                               (stats['oddR'] >= MIN_ODDR) &\n",
    "                          (stats['adjPval']<=pvalue) & \n",
    "                          (stats['ESB_test']>=0.01)]\n",
    "        positive_positions |= set(stats_high['start_loc'].tolist())\n",
    "    \n",
    "    modfracs=[]\n",
    "    for file, name in zip(filelist, samples):\n",
    "        stats = pd.read_csv(file, sep = '\\t')\n",
    "        stats['sample'] = name\n",
    "        modfracs.append(stats[stats['start_loc'].isin(positive_positions)])\n",
    "        \n",
    "    \n",
    "    if len(modfracs) > 0:\n",
    "        ret = pd.concat(modfracs).sort_values(by=['start_loc', 'sample']).reset_index(drop=True)\n",
    "        return ret\n",
    "\n",
    "def aggregate_adjacent_sites(modfracs, max_distance, minimum_coverage):\n",
    "    modsites_raw = sorted(modfracs['start_loc'].unique())\n",
    "    continuity = np.diff(modsites_raw) <= max_distance\n",
    "\n",
    "    modsite_groups = [[modsites_raw.pop(0)]]\n",
    "    for pos, cont in zip(modsites_raw, continuity):\n",
    "        if cont:\n",
    "            modsite_groups[-1].append(pos)\n",
    "        else:\n",
    "            modsite_groups.append([pos])\n",
    "\n",
    "    purified = []\n",
    "    for modsites in modsite_groups:\n",
    "        modfracsites = modfracs[modfracs['start_loc'].isin(modsites)]\n",
    "\n",
    "        modfracsites_high = modfracsites[modfracsites['total_reads'] >= minimum_coverage]\n",
    "        maxpos = modfracsites.loc[modfracsites_high['oddR'].idxmax()]\n",
    "        purified.append(modfracsites[modfracsites['start_loc'] == maxpos['start_loc']])\n",
    "\n",
    "    return pd.concat(purified)\n",
    "\n",
    "def identify_position(p, anno):\n",
    "    matching = anno[(anno['start'] <= p) & (p < anno['end'])]\n",
    "    return(matching.index.tolist()[0])\n",
    "\n",
    "modfracs = load_combine_all(\n",
    "    [f'eligos2/pairdiff4/{name}/combine.txt' for name in sample_names],\n",
    "              sample_names)\n",
    "\n",
    "eligos2_freqmods = get_modified_sites(\n",
    "    [f'eligos2/pairdiff4/{name}/combine.txt' for name in sample_names],\n",
    "              sample_names, pvalue=0.001)\n",
    "\n",
    "eligos2_psites = aggregate_adjacent_sites(eligos2_freqmods, AGG_DIS, MIN_COV)\n",
    "\n",
    "def summarize_into_table(aggrtbl):\n",
    "    mod_frac_all = pd.pivot_table(aggrtbl, columns='sample', values=['oddR'],\n",
    "                                 index=['start_loc', 'strand']).fillna(0)['oddR'].copy()\n",
    "    coverage = pd.pivot_table(aggrtbl, columns='sample', values=['total_reads'],\n",
    "                             index=['start_loc', 'strand']).fillna(0)['total_reads'].copy()\n",
    "    esb = pd.pivot_table(aggrtbl, columns='sample', values=['ESB_test'],\n",
    "                             index=['start_loc', 'strand']).fillna(0)['ESB_test'].copy()\n",
    "    return esb, mod_frac_all, coverage\n",
    "\n",
    "def add_context_info(sitemodfrac, refseq, flanksize):\n",
    "    context_list=[]\n",
    "    for pos, strand in sitemodfrac.index:\n",
    "\n",
    "        context=refseq[pos-flanksize:pos].lower() + refseq[pos:pos+1].upper()+ refseq[pos+1:pos+1+flanksize].lower()\n",
    "        if strand =='-':\n",
    "            context = str(Seq.Seq(context).reverse_complement())\n",
    "        else:\n",
    "            context = context\n",
    "        context_list.append(context)\n",
    "        \n",
    "    return(context_list)\n",
    "\n",
    "\n",
    "eligos2_esb, eligos2_sitemodfrac, eligos2_sitecov= summarize_into_table(eligos2_psites)\n",
    "\n",
    "\n",
    "#eligos2_sitemodfrac['lnc'] = eligos2_sitemodfrac.index.to_series().apply(partial(identify_position, lncanno=lncanno))\n",
    "eligos2_sitemodfrac['context'] = add_context_info(eligos2_sitemodfrac, refseq, 10)\n",
    "eligos2_sitemodfrac = eligos2_sitemodfrac.reset_index().set_index('start_loc')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_position(p, strand, anno):\n",
    "    anno = anno[anno['strand']==strand]\n",
    "    matching = anno[(anno['start'] <= p) & (p < anno['end'])]\n",
    "    try:\n",
    "        return(matching['name'].tolist()[0])\n",
    "    except:\n",
    "        return('no_match')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "eligos2_sitemodfrac['gene'] = [identify_position(p, line['strand'], target_anno) for p, line in eligos2_sitemodfrac.iterrows()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>strand</th>\n",
       "      <th>latent_lytic</th>\n",
       "      <th>latent_react</th>\n",
       "      <th>lytic_latent</th>\n",
       "      <th>lytic_react</th>\n",
       "      <th>react_latent</th>\n",
       "      <th>react_lytic</th>\n",
       "      <th>context</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_loc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2361</th>\n",
       "      <td>-</td>\n",
       "      <td>0.421196</td>\n",
       "      <td>0.600544</td>\n",
       "      <td>2.374193</td>\n",
       "      <td>1.425772</td>\n",
       "      <td>1.665156</td>\n",
       "      <td>0.701374</td>\n",
       "      <td>guuugcaugcAuuugccgugu</td>\n",
       "      <td>RNA2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>-</td>\n",
       "      <td>0.677533</td>\n",
       "      <td>0.923316</td>\n",
       "      <td>2.353971</td>\n",
       "      <td>2.066408</td>\n",
       "      <td>1.216255</td>\n",
       "      <td>0.737841</td>\n",
       "      <td>gaguaguagaCacaauaaaug</td>\n",
       "      <td>RNA2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>-</td>\n",
       "      <td>0.709328</td>\n",
       "      <td>1.176367</td>\n",
       "      <td>2.134932</td>\n",
       "      <td>2.511638</td>\n",
       "      <td>1.196994</td>\n",
       "      <td>0.775596</td>\n",
       "      <td>uagcaacacaGaguaguagac</td>\n",
       "      <td>RNA2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>-</td>\n",
       "      <td>2.066086</td>\n",
       "      <td>1.670607</td>\n",
       "      <td>0.871557</td>\n",
       "      <td>0.923196</td>\n",
       "      <td>0.944082</td>\n",
       "      <td>1.236738</td>\n",
       "      <td>uuuuuuuuuuCuuugccuuca</td>\n",
       "      <td>RNA2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2478</th>\n",
       "      <td>-</td>\n",
       "      <td>0.587052</td>\n",
       "      <td>0.631269</td>\n",
       "      <td>2.082594</td>\n",
       "      <td>1.198116</td>\n",
       "      <td>1.738188</td>\n",
       "      <td>0.929985</td>\n",
       "      <td>ucggucgcacGgauuauuccu</td>\n",
       "      <td>RNA2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212675</th>\n",
       "      <td>-</td>\n",
       "      <td>0.346109</td>\n",
       "      <td>0.268048</td>\n",
       "      <td>5.343878</td>\n",
       "      <td>0.784762</td>\n",
       "      <td>10.609534</td>\n",
       "      <td>2.495068</td>\n",
       "      <td>guuucgccgaGuuucgcccac</td>\n",
       "      <td>US18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212745</th>\n",
       "      <td>-</td>\n",
       "      <td>0.779059</td>\n",
       "      <td>0.439208</td>\n",
       "      <td>1.374584</td>\n",
       "      <td>0.563779</td>\n",
       "      <td>2.636093</td>\n",
       "      <td>2.025575</td>\n",
       "      <td>cagugcgccaAcggcauagug</td>\n",
       "      <td>US18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212879</th>\n",
       "      <td>-</td>\n",
       "      <td>0.331733</td>\n",
       "      <td>0.721330</td>\n",
       "      <td>3.425446</td>\n",
       "      <td>2.173948</td>\n",
       "      <td>1.951809</td>\n",
       "      <td>0.633254</td>\n",
       "      <td>ccgaacagcaGcucuuccaau</td>\n",
       "      <td>US18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212918</th>\n",
       "      <td>-</td>\n",
       "      <td>0.375777</td>\n",
       "      <td>0.488470</td>\n",
       "      <td>3.727199</td>\n",
       "      <td>1.682376</td>\n",
       "      <td>2.218336</td>\n",
       "      <td>0.807710</td>\n",
       "      <td>ucaccaucguGccgcuccauc</td>\n",
       "      <td>US18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212955</th>\n",
       "      <td>-</td>\n",
       "      <td>0.598073</td>\n",
       "      <td>0.867147</td>\n",
       "      <td>2.639978</td>\n",
       "      <td>2.229344</td>\n",
       "      <td>1.740609</td>\n",
       "      <td>0.735659</td>\n",
       "      <td>accgccucggUuuccgaacac</td>\n",
       "      <td>US18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample    strand  latent_lytic  latent_react  lytic_latent  lytic_react  \\\n",
       "start_loc                                                                 \n",
       "2361           -      0.421196      0.600544      2.374193     1.425772   \n",
       "2386           -      0.677533      0.923316      2.353971     2.066408   \n",
       "2396           -      0.709328      1.176367      2.134932     2.511638   \n",
       "2433           -      2.066086      1.670607      0.871557     0.923196   \n",
       "2478           -      0.587052      0.631269      2.082594     1.198116   \n",
       "...          ...           ...           ...           ...          ...   \n",
       "212675         -      0.346109      0.268048      5.343878     0.784762   \n",
       "212745         -      0.779059      0.439208      1.374584     0.563779   \n",
       "212879         -      0.331733      0.721330      3.425446     2.173948   \n",
       "212918         -      0.375777      0.488470      3.727199     1.682376   \n",
       "212955         -      0.598073      0.867147      2.639978     2.229344   \n",
       "\n",
       "sample     react_latent  react_lytic                context    gene  \n",
       "start_loc                                                            \n",
       "2361           1.665156     0.701374  guuugcaugcAuuugccgugu  RNA2.7  \n",
       "2386           1.216255     0.737841  gaguaguagaCacaauaaaug  RNA2.7  \n",
       "2396           1.196994     0.775596  uagcaacacaGaguaguagac  RNA2.7  \n",
       "2433           0.944082     1.236738  uuuuuuuuuuCuuugccuuca  RNA2.7  \n",
       "2478           1.738188     0.929985  ucggucgcacGgauuauuccu  RNA2.7  \n",
       "...                 ...          ...                    ...     ...  \n",
       "212675        10.609534     2.495068  guuucgccgaGuuucgcccac    US18  \n",
       "212745         2.636093     2.025575  cagugcgccaAcggcauagug    US18  \n",
       "212879         1.951809     0.633254  ccgaacagcaGcucuuccaau    US18  \n",
       "212918         2.218336     0.807710  ucaccaucguGccgcuccauc    US18  \n",
       "212955         1.740609     0.735659  accgccucggUuuccgaacac    US18  \n",
       "\n",
       "[462 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eligos2_filtered = eligos2_sitemodfrac[eligos2_sitemodfrac['gene']!='no_match']\n",
    "eligos2_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(eligos2_filtered['gene']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/blaze/ari/conda/envs/lab/lib/python3.8/site-packages/pandas/core/frame.py:5238: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n"
     ]
    }
   ],
   "source": [
    "eligos2_filtered.replace([np.inf, -np.inf], np.nan, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2528373/2185817620.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  eligos2_filtered['pos'] = poslist\n"
     ]
    }
   ],
   "source": [
    "poslist = []\n",
    "for _, line in eligos2_filtered.iterrows():\n",
    "    gene = line['gene']\n",
    "    anno = target_anno[target_anno['name']==gene]\n",
    "    if len(anno) == 1:\n",
    "        if line['strand'] == '+':\n",
    "            \n",
    "\n",
    "            pos = _ - list(anno['start'])[0]\n",
    "            \n",
    "        else:\n",
    "            pos = list(anno['end'])[0] - _\n",
    "    else:\n",
    "        if line['strand'] == '+':\n",
    "            anno = anno.sort_values(by='start')\n",
    "            pos = 0\n",
    "         \n",
    "            for s, e in np.array(anno[['start', 'end']]):\n",
    "\n",
    "                if _ in range(s,e):\n",
    "                    pos += _ - s\n",
    "                else:\n",
    "                    pos += e-s\n",
    "        else:\n",
    "            anno = anno.sort_values(by='end')\n",
    "            pos = 0\n",
    "            for s, e in np.array(anno[['start', 'end']]):\n",
    "                if _ in range(s,e):\n",
    "                    pos += e - _\n",
    "                else:\n",
    "                    pos += e-s\n",
    "    poslist.append(pos)\n",
    "    \n",
    "eligos2_filtered['pos'] = poslist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2528373/2429560800.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  eligos2_filtered['pos_stride'] = np.floor(eligos2_filtered['pos']/STRIDESIZE).astype(int) * STRIDESIZE\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "98"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STRIDESIZE = 300\n",
    "eligos2_filtered['pos_stride'] = np.floor(eligos2_filtered['pos']/STRIDESIZE).astype(int) * STRIDESIZE\n",
    "\n",
    "eligos2_stride = pd.pivot_table(eligos2_filtered, index=['gene', 'pos_stride'], values= sample_names,\n",
    "       aggfunc='median')\n",
    "len(eligos2_stride)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "eligos2_stride['start'] = [target_anno[target_anno['name']==g[0]]['start'].to_list()[0] for g in eligos2_stride.index]\n",
    "eligos2_stride['stride'] = [g[1] for g in eligos2_stride.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>latent_lytic</th>\n",
       "      <th>latent_react</th>\n",
       "      <th>lytic_latent</th>\n",
       "      <th>lytic_react</th>\n",
       "      <th>react_latent</th>\n",
       "      <th>react_lytic</th>\n",
       "      <th>start</th>\n",
       "      <th>stride</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>pos_stride</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">RNA2.7</th>\n",
       "      <th>0</th>\n",
       "      <td>0.644357</td>\n",
       "      <td>0.658208</td>\n",
       "      <td>2.021157</td>\n",
       "      <td>1.141107</td>\n",
       "      <td>2.286934</td>\n",
       "      <td>1.113305</td>\n",
       "      <td>2361</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0.703033</td>\n",
       "      <td>0.682544</td>\n",
       "      <td>1.840351</td>\n",
       "      <td>1.016111</td>\n",
       "      <td>2.078408</td>\n",
       "      <td>1.196373</td>\n",
       "      <td>2361</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.778404</td>\n",
       "      <td>0.590458</td>\n",
       "      <td>1.631523</td>\n",
       "      <td>0.837502</td>\n",
       "      <td>2.188478</td>\n",
       "      <td>2.006540</td>\n",
       "      <td>2361</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900</th>\n",
       "      <td>0.655370</td>\n",
       "      <td>0.607624</td>\n",
       "      <td>1.754831</td>\n",
       "      <td>0.872747</td>\n",
       "      <td>2.424549</td>\n",
       "      <td>1.462615</td>\n",
       "      <td>2361</td>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>0.724244</td>\n",
       "      <td>0.649639</td>\n",
       "      <td>1.915389</td>\n",
       "      <td>0.926569</td>\n",
       "      <td>2.442073</td>\n",
       "      <td>1.333524</td>\n",
       "      <td>2361</td>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">US12</th>\n",
       "      <th>300</th>\n",
       "      <td>0.531502</td>\n",
       "      <td>0.500956</td>\n",
       "      <td>2.346576</td>\n",
       "      <td>0.983949</td>\n",
       "      <td>2.735875</td>\n",
       "      <td>1.168692</td>\n",
       "      <td>206323</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.700813</td>\n",
       "      <td>0.835082</td>\n",
       "      <td>2.007207</td>\n",
       "      <td>1.314658</td>\n",
       "      <td>1.884192</td>\n",
       "      <td>1.252344</td>\n",
       "      <td>206323</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">US18</th>\n",
       "      <th>0</th>\n",
       "      <td>0.486925</td>\n",
       "      <td>0.604900</td>\n",
       "      <td>3.032712</td>\n",
       "      <td>1.928162</td>\n",
       "      <td>2.085073</td>\n",
       "      <td>0.771684</td>\n",
       "      <td>212150</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0.564665</td>\n",
       "      <td>0.481136</td>\n",
       "      <td>2.551145</td>\n",
       "      <td>0.768867</td>\n",
       "      <td>3.063975</td>\n",
       "      <td>1.637901</td>\n",
       "      <td>212150</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.706424</td>\n",
       "      <td>0.540813</td>\n",
       "      <td>1.826257</td>\n",
       "      <td>0.964348</td>\n",
       "      <td>2.167300</td>\n",
       "      <td>1.494512</td>\n",
       "      <td>212150</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample             latent_lytic  latent_react  lytic_latent  lytic_react  \\\n",
       "gene   pos_stride                                                          \n",
       "RNA2.7 0               0.644357      0.658208      2.021157     1.141107   \n",
       "       300             0.703033      0.682544      1.840351     1.016111   \n",
       "       600             0.778404      0.590458      1.631523     0.837502   \n",
       "       900             0.655370      0.607624      1.754831     0.872747   \n",
       "       1200            0.724244      0.649639      1.915389     0.926569   \n",
       "...                         ...           ...           ...          ...   \n",
       "US12   300             0.531502      0.500956      2.346576     0.983949   \n",
       "       600             0.700813      0.835082      2.007207     1.314658   \n",
       "US18   0               0.486925      0.604900      3.032712     1.928162   \n",
       "       300             0.564665      0.481136      2.551145     0.768867   \n",
       "       600             0.706424      0.540813      1.826257     0.964348   \n",
       "\n",
       "sample             react_latent  react_lytic   start  stride  \n",
       "gene   pos_stride                                             \n",
       "RNA2.7 0               2.286934     1.113305    2361       0  \n",
       "       300             2.078408     1.196373    2361     300  \n",
       "       600             2.188478     2.006540    2361     600  \n",
       "       900             2.424549     1.462615    2361     900  \n",
       "       1200            2.442073     1.333524    2361    1200  \n",
       "...                         ...          ...     ...     ...  \n",
       "US12   300             2.735875     1.168692  206323     300  \n",
       "       600             1.884192     1.252344  206323     600  \n",
       "US18   0               2.085073     0.771684  212150       0  \n",
       "       300             3.063975     1.637901  212150     300  \n",
       "       600             2.167300     1.494512  212150     600  \n",
       "\n",
       "[98 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eligos2_stride = eligos2_stride.sort_values(by=['start', 'stride'])\n",
    "eligos2_stride"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2528373/2724888119.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  target_anno['length'] = target_anno['end'] - target_anno['start']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'length': {'RNA1.2': 1051.0,\n",
       "  'RNA2.7': 2513.0,\n",
       "  'RNA4.9': 4896.0,\n",
       "  'UL112': 1027.5,\n",
       "  'UL122': 581.0,\n",
       "  'UL132': 813.0,\n",
       "  'UL141': 1017.0,\n",
       "  'UL145': 393.0,\n",
       "  'UL17': 315.0,\n",
       "  'UL22A': 159.0,\n",
       "  'UL25': 1971.0,\n",
       "  'UL34': 1224.0,\n",
       "  'UL4': 453.0,\n",
       "  'UL40': 666.0,\n",
       "  'UL48': 6729.0,\n",
       "  'UL48A': 228.0,\n",
       "  'UL5': 501.0,\n",
       "  'UL53': 1131.0,\n",
       "  'UL54': 3729.0,\n",
       "  'UL69': 2220.0,\n",
       "  'UL73': 408.0,\n",
       "  'UL82': 1683.0,\n",
       "  'UL84': 1761.0,\n",
       "  'UL89': 1012.5,\n",
       "  'UL99': 573.0,\n",
       "  'US1': 471.0,\n",
       "  'US12': 846.0,\n",
       "  'US18': 825.0}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_anno['length'] = target_anno['end'] - target_anno['start']\n",
    "target_length = pd.pivot_table(target_anno, index='name', values='length',)\n",
    "target_length.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "yticks = []\n",
    "for ind in eligos2_stride.index:\n",
    "    gene = ind[0]\n",
    "    end = ind[1]+300\n",
    "    if end > target_length.to_dict()['length'][gene]:\n",
    "        end = target_length.to_dict()['length'][gene]\n",
    "    yticks.append(f'{ind[0]}:{ind[1]+1}:{int(end)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
